src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,324 @@
+    ARG_CHECK(adaptor != NULL);
+    ARG_CHECK(presig65 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    /* P */
+    ret &= !!secp256k1_xonly_pubkey_load(ctx, &pk, pubkey);
+
+    /* s0 */
+    secp256k1_scalar_set_b32(&s0, &presig65[33], &overflow);
+    ret &= !overflow;
+
+    /* R0 */
+    ret &= !!secp256k1_xonly_pubkey_parse(ctx, &pkr0, &presig65[1]);
+    ret &= !!secp256k1_xonly_pubkey_load(ctx, &r0, &pkr0);
+
+    /* Compute e */
+    secp256k1_fe_get_b32(buf, &pk.x);
+    secp256k1_schnorrsig_challenge(&e, &presig65[1], msg32, msglen, buf);
+
+    /* Compute rj = s0*G + (-e) * pkj */
+    secp256k1_scalar_negate(&e, &e);
+    secp256k1_gej_set_ge(&pkj, &pk);
+    secp256k1_ecmult(&rj, &pkj, &e, &s0);
+
+    /* R */
+    secp256k1_ge_set_gej_var(&r, &rj);
+    ret &= !secp256k1_ge_is_infinity(&r);
+
+    /* T = R0 + (- R) */
+    secp256k1_gej_neg(&rj, &rj);
+    secp256k1_gej_add_ge_var(&tj, &rj, &r0, NULL);
+    if (presig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        ;
+    } else if (presig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_gej_neg(&tj, &tj);
+    } else {
+        ret = 0;
+    }
+    secp256k1_ge_set_gej(&t, &tj);
+    ret &= !!secp256k1_eckey_pubkey_serialize(&t, adaptor, &size, 1);
+
+    secp256k1_memczero(adaptor, 33, !ret);
+
+    return ret;
+}
+
+int secp256k1_schnorr_adaptor_adapt(const secp256k1_context *ctx, unsigned char *sig64, const unsigned char *presig65, const unsigned char *secadaptor) {
+    secp256k1_scalar s0;
+    secp256k1_scalar s;
+    secp256k1_scalar t;
+    int overflow;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(presig65 != NULL);
+    ARG_CHECK(presig65[0] == SECP256K1_TAG_PUBKEY_EVEN || presig65[0] == SECP256K1_TAG_PUBKEY_ODD);
+    ARG_CHECK(secadaptor != NULL);
+
+    /* s0 */
+    secp256k1_scalar_set_b32(&s0, &presig65[33], &overflow);
+    ret &= !overflow;
+
+    /* t */
+    secp256k1_scalar_set_b32(&t, secadaptor, &overflow);
+    ret &= !overflow;
+
+    if (presig65[0] == SECP256K1_TAG_PUBKEY_EVEN) {
+        secp256k1_scalar_add(&s, &s0, &t);
+    } else if (presig65[0] == SECP256K1_TAG_PUBKEY_ODD) {
+        secp256k1_scalar_negate(&t, &t);
+        secp256k1_scalar_add(&s, &s0, &t);
+    }
+
+    memcpy(sig64, &presig65[1], 32);
+    secp256k1_scalar_get_b32(&sig64[32], &s);
+    secp256k1_memczero(sig64, 64, !ret);
+    secp256k1_scalar_clear(&s);
+    secp256k1_scalar_clear(&t);
+
+    return ret;
+}
+
+int secp256k1_schnorr_adaptor_extract_sec(const secp256k1_context *ctx, unsigned char *secadaptor, const unsigned char *presig65, const unsigned char *sig64) {
+    secp256k1_scalar s0;
+    secp256k1_scalar s;
+    secp256k1_scalar t;
+    int overflow;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secadaptor != NULL);
+    ARG_CHECK(presig65 != NULL);
+    ARG_CHECK(presig65[0] == SECP256K1_TAG_PUBKEY_EVEN || presig65[0] == SECP256K1_TAG_PUBKEY_ODD);
```

Comment:
I second this. libsecp handles other scenarios similarly:
- If the nonce is zero, `shcnorr_sign` sets `ret` to zero and continues with signature creation instead of calling an illegal callback.
- `eckey_pubkey_parse` also returns 0 for an invalid first byte.

This change also needs to be applied to the `schnorr_adaptor_adapt` API. 

At present, [our ctime test](https://github.com/BlockstreamResearch/secp256k1-zkp/pull/268#issuecomment-1694655543) considers the `adapt` and `extract_secadaptor` APIs to be non-const-time wrt `presig`. Therefore, it is safe to return 0 immediately. However, if we plan on making these APIs const-time wrt `presig`, we should set `ret` to zero instead.